### PR TITLE
My take on this week's homework

### PR DESCRIPTION
**Pull request by CapnKernel**
Hi!

This is where I'm at with the homework.

I'm a bit concerned that the README.md file implies multiple models for normalisation.  The only one that makes sense to me is to pull out client.  Can you please suggest additional targets for normalisation?

Thanks!

### CODE REVIEW

File: timetracker/entries/migrations/0003_move_project_client_to_model.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    # see if there's already a corresponding Client.  If not, create it.
+    # Then set the client property to be the new or existing Client.
+    for project in Project.objects.all():
+        # Is there an existing client of this name?
+        try:
+            c = Client.objects.get(name=project.client_as_str)
+        except Client.DoesNotExist:
+            # No, so create it
+            c = Client(name=project.client_as_str)
+            c.save()
```
Comment (funkybob) [inline timetracker/entries/migrations/0003_move_project_client_to_model.py]: These two steps could have been one using Client.objects.create()

File: timetracker/entries/migrations/0003_move_project_client_to_model.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+def rewrite_client_str_to_client_model(apps, schema_editor):
+    Client = apps.get_model("entries", "Client")
+    Project = apps.get_model("entries", "Project")
+    
+    # For each project, get the text version of the client name, and
+    # see if there's already a corresponding Client.  If not, create it.
+    # Then set the client property to be the new or existing Client.
+    for project in Project.objects.all():
+        # Is there an existing client of this name?
+        try:
```
Comment (funkybob) [inline timetracker/entries/migrations/0003_move_project_client_to_model.py]: This try/except could have been done using Client.objects.get_or_create()

File: README.md [hunk 3]
```diff
@@ -10,12 +10,12 @@ https://melbdjango.github.io/lesson-two/
 
 If you did the homework [last week][lesson-one], you should already have a virtualenv that you can use this week.
 
-- [ ] [Fork this repository][gh-fork]
-- [ ] Clone the repo to your own machine
-- [ ] Normalise the model we created in class by creating new [Django models][dj-models] that better express the
+- [X] [Fork this repository][gh-fork]
```
Comment (sesh) [inline README.md:13]: Checking things off!

File: timetracker/entries/migrations/0003_move_project_client_to_model.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    dependencies = [
+        ('entries', '0002_auto_20150723_0819'),
+    ]
+
+    # Switch from a charfield-based client, to a model-based one.
+    operations = [
+        migrations.AlterField(
+            model_name='project',
+            name='client',
+            field=models.CharField(max_length=200, null=True),
```
Comment (sesh) [inline timetracker/entries/migrations/0003_move_project_client_to_model.py:44]: Is there a specific reason that you're making this field nullable?

File: timetracker/entries/migrations/0004_add_entry_pluralisation.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('entries', '0003_move_project_client_to_model'),
+    ]
+
+    operations = [
+        migrations.AlterModelOptions(
+            name='entry',
+            options={'verbose_name_plural': 'entries'},
```
Comment (sesh) [inline timetracker/entries/migrations/0004_add_entry_pluralisation.py:16]: Even more bonus points!

File: timetracker/entries/models.py [hunk 6]
```diff
@@ -2,24 +2,40 @@
...
+    name = models.CharField(max_length=200)
 
+    def __str__(self):
+        return '<Client: {}>'.format(self.name)
+
+    
 class Project(models.Model):
-    client = models.CharField(max_length=200)
     name = models.CharField(max_length=200)
+    client = models.ForeignKey(Client)
```
Comment (sesh) [inline timetracker/entries/models.py:15]: Little heads up: it's safer to put the model name in quotes in `ForeignKey` and `ManyToMany` field definitions.

**Comment by sesh** [comment]
Nice work @CapnKernel.

**Pull request closed without merging**